README.md
# ff_contact_avatar
A highly configurable Column widget for displaying a contact avatar with name and short message.  Library also includes the ability the create a ListView using the FFContactAvatarList Widget or using the convenience functions `fromModelList` (to create a ListView directly, suited to a small number or items) and `fromModelListWithBuilder` (to create a ListView using the .builder approach, suited to a large number of items).


![](demo.png)

## Add dependency
```
dependencies:
  ff_contact_avatar: ^0.1.1
```

## Basic use (Single Avatar item)
To manually configure each item, simply create FFContactAvatar objects using the approach below...
```
import 'package:flutter/material.dart';
import 'package:ff_contact_avatar/ff_contact_avatar.dart';
...
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            FFContactAvatar(
              name: 'Ed Roberts',
              message: 'He\'s a man',
              showBadge: true,
            ),
            SizedBox(height: 10),
            FFContactAvatar(
              name: 'Mr Ed',
              message: 'He\'s a horse',
              image: Image.network(
                  '<URL of file here or swap for Image.asset(...)'),
              showBadge: true,
            ),
          ],
        ),
      ),
    );
```

## Basic use with Theme
To use a specific theme for each item, specifying it using the inline option below...
```
...
            FFContactAvatar(
              name: 'Ed Roberts',
              message: 'He\'s a man',
              showBadge: true,
              theme: FFContactAvatarTheme(
                backgroundColor: Colors.blueAccent,
                avatarElevation: 20.0,
              ),
            ),
...
```

## Simple ListView creation
If you need a list for only a small number of items, it is recommended to create a list of FFContactAvatarModel objects to store the data, before creating the ListView directly from the list of models...
```
...
final List<FFContactAvatarModel> models = [
    FFContactAvatarModel(
      name: 'Ed Roberts',
      message: 'He\s a man',
      showBadge: true,
    ),
    FFContactAvatarModel(
      name: 'Mr Ed',
      message: 'He\s a horse',
      showBadge: false,
      image: Image.network(
          'https://upload.wikimedia.org/wikipedia/commons/8/81/Mister_Ed.JPG'),
    ),
    FFContactAvatarModel(
      name: 'Mrs Roberts',
      message: 'Ed\'s Wife',
    ),
    FFContactAvatarModel(
      name: 'Mrs Horse',
      message: 'Also Ed\'s Wife',
      showBadge: true,
    ),
  ];
...
return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            // Output a ListView using the manual data set (NO builder)
            Text('Sample using full ListView'),
            Expanded(
              child: FFContactAvatarList.fromModelList(
                models,
                scrollDirection: Axis.horizontal,
                onSelectAvatar: (index) {
                  print('Index ' + index.toString() + ' selected.');
                },
              ),
            ),
            SizedBox(
              height: 1,
              child: Container(
                color: Colors.grey,
              ),
            ),
          ],
        ),
      ),
    );
...
```

## Creating a ListView Builder 
Using ListView.builder approach is recommended when you need to display a large number of items in a ListView.  This approach allows the builder to handle displaying only those items that need to be visible to the user, deriving better performance.
```
//
// Create the list of data as before ... it just needs to be a List<FFContactAvatarModel> object
//
return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            // Output a ListView using the .builder approach (using long list of generated data)
            Text('Sample using ListView.Builder'),
            Expanded(
              child: FFContactAvatarList.fromModelListWithBuilder(
                longList,
                context: context,
                scrollDirection: Axis.horizontal,
                onSelectAvatar: (index) {
                  print('Index ' + index.toString() + ' selected.');
                },
              ),
            ),
          ],
        ),
      ),
    );
``` 

## Theme
The avatar has a Theme class which can be used to define the majority of appearance settings for the avatar.
* backgroundColor: The background color for the CircleAvatar when initials are being displayed (i.e. no image) (default = blueGrey)
* foregroundColor: The color to use for the text of initials (default = white)
* badgeColor: The color to use for the notification badge circle (default = red)
* avatarRadius: The radius of the CircleAvatar widget (default = 31)
* initialsTextStyle: The default text style for initials (default = bold, 26pt with 1.4 letter spacing)
* nameTextStyle: The default text style for the name (default = bold, 13pt)
* messageTextStyle: The default text style for the message (default = normal, 11pt)
* verticalPadding: The space to include between each vertical component (default = 2.0)
* avatarElevation: The elevation to be used for drop shadow effect (default = 4.0)
* itemExtent: The extent (cross-axis width/height) for each item (only applies to ListView) (default = 80)
* listPadding: The padding (as an EdgeInsets object) to use in the ListView (default = EdgeInsets.all(16.0))

## FFContactAvatarModel
The Model object is provided to encapsulate the creation of data in a way that is compatible with the easy creation of FFContactAvatar and related List Widgets.

The following parameters are available:
```
  final String name;        // Name value 
  final String message;     // Message value
  final Image image;        // Optional image value
  final bool showBadge;     // Bool value (should notification badge be displayed, default = false
  final VoidCallback onTap; // onTap handler (void callback)
```

## FFContactAvatar Attributes
* theme: A FFContactAvatarTheme theme object
* image: An Image object for the image to display in the CircleAvatar
* name: A String of the contact name to display (will also be used to create the initials if no image is provided, with a maximum initials length of 2)
* message: A string of the message to display
* showBadge: A bool value indicating if the notification badge should be displayed (default = false)
* onTap: A void callback function triggered when the circle avatar is tapped
* model: A FFContactAvatarModel object which can be used to provide the name, message, image and showBadge options instead of specifying individually.

## FFContactAvatarList Attributes
* models: A List<FFContactAvatarModel> object which provides the data to use.
* scrollDirection: An Axis attribute which specifies the direction of the ListView (e.g. Axis.horizontal)
* theme: An FFContactAvatarTheme object which can specific the theme of ALL items in the ListView
* onSelectAvatar: A callback function (which requires a int type index parameter) that will be called when a child of the ListView is tapped, with the index representing the zero-based index of the selected item. 

## FFContactAvatarList Convenience Functions
The static function `fromModelList` can be called to easily create a ListView using standard theme and options...
```
static StatelessWidget fromModelList(
    List<FFContactAvatarModel> models, {
    Axis scrollDirection = Axis.horizontal,
    FFContactAvatarListSelected onSelectAvatar,
  })
```

The static function `fromModelListWithBuilder` can be called to easily create a ListViewBuilder using standard theme and options...
```
static StatelessWidget fromModelListWithBuilder(
    List<FFContactAvatarModel> models, {
    Axis scrollDirection = Axis.horizontal,
    FFContactAvatarListSelected onSelectAvatar,
    @required BuildContext context,
  })
```

CHANGELOG.md
## [0.1.1] - 30-NOV-2019

* Added FFContactAvatarModel as a class to pass in all data attributes
* Support instantiating FFContactAvatar using a model class (optional instead of passing individual elements)
* Added FFContactAvatarList to support creating ListView for a list of model objects
* Added onTap to provide a callback function to FFContactAvatar


## [0.1.0] - 27-NOV-2019

* First release of FFContactAvatar


lib/ff_contact_avatar.dart
library flutter_ff_contact_avatar;

import 'package:flutter/material.dart';
import 'dart:math' show min;

import 'ff_contact_avatar_model.dart';
import 'package:ff_contact_avatar/ff-contact-avatar-theme.dart';

export 'package:ff_contact_avatar/ff_contact_avatar_list.dart';
export 'package:ff_contact_avatar/ff-contact-avatar-theme.dart';
export 'ff_contact_avatar_model.dart';

// ignore: must_be_immutable
class FFContactAvatar extends StatelessWidget {
  Image image;
  String name;
  String message;
  bool showBadge;
  final FFContactAvatarModel model;
  final VoidCallback onTap;

  FFContactAvatarTheme theme;

  FFContactAvatar({
    FFContactAvatarTheme theme,
    this.model,
    this.image,
    this.name,
    this.message,
    this.showBadge = false,
    this.onTap,
  }) {
    this.theme = (theme == null) ? FFContactAvatarTheme.defaultTheme : theme;

    if (model != null) {
      this.image = model.image;
      this.name = model.name;
      this.message = model.message;
      this.showBadge = model.showBadge;
    }
  }

  static const double avatarRadius = 31;

  String _getInitials() {
    var nameParts = name.split(" ").map((elem) {
      return elem[0];
    });

    if (nameParts.length == 0) {
      return "";
    }

    int numberOfParts = min(2, nameParts.length);
    return nameParts.join().substring(0, numberOfParts);
  }

  CircleAvatar _makeImageAvatar() {
    return CircleAvatar(
      radius: avatarRadius,
      backgroundImage: this.image.image,
    );
  }

  CircleAvatar _makeInitialsAvatar() {
    return CircleAvatar(
      radius: avatarRadius,
      child: Text(
        _getInitials(),
        style: theme.initialsTextStyle,
      ),
      backgroundColor: theme.backgroundColor,
      foregroundColor: theme.foregroundColor,
    );
  }

  Widget _makeCircleAvatar() {
    CircleAvatar ca =
        (image != null) ? _makeImageAvatar() : _makeInitialsAvatar();

    Color badgeColor = showBadge ? theme.badgeColor : Colors.transparent;

    return GestureDetector(
      onTap: this.onTap,
      child: Material(
        elevation: theme.avatarElevation,
        borderRadius: BorderRadius.circular(avatarRadius),
        child: Stack(
          children: <Widget>[
            ca,
            Positioned(
              top: 1,
              right: 1,
              width: 16,
              height: 16,
              child: Container(
                decoration: BoxDecoration(
                  color: badgeColor,
                  borderRadius: BorderRadius.circular(8.0),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  SizedBox _makePadding() {
    return SizedBox(
      height: theme.verticalPadding,
    );
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        _makeCircleAvatar(),
        _makePadding(),
        Text(
          name,
          style: theme.nameTextStyle,
        ),
        _makePadding(),
        Text(
          message,
          style: theme.messageTextStyle,
        ),
      ],
    );
  }
}


lib/ff_contact_avatar_list.dart
import 'package:flutter/material.dart';
import 'ff_contact_avatar.dart';
import 'ff_contact_avatar_model.dart';

typedef void FFContactAvatarListSelected(int index);

class FFContactAvatarList extends StatelessWidget {
  final List<FFContactAvatarModel> models;
  final Axis scrollDirection;
  final FFContactAvatarTheme theme;
  final FFContactAvatarListSelected onSelectAvatar;

  FFContactAvatarList({
    this.models,
    this.scrollDirection,
    this.theme,
    this.onSelectAvatar,
  });

  static StatelessWidget fromModelList(
    List<FFContactAvatarModel> models, {
    Axis scrollDirection = Axis.horizontal,
    FFContactAvatarListSelected onSelectAvatar,
  }) {
    return FFContactAvatarList(
      models: models,
      scrollDirection: scrollDirection,
      theme: FFContactAvatarTheme.defaultTheme,
      onSelectAvatar: onSelectAvatar,
    );
  }

  static StatelessWidget fromModelListWithBuilder(
    List<FFContactAvatarModel> models, {
    Axis scrollDirection = Axis.horizontal,
    FFContactAvatarListSelected onSelectAvatar,
    @required BuildContext context,
  }) {
    FFContactAvatarTheme theme = FFContactAvatarTheme.defaultTheme;

    return ListView.builder(
      scrollDirection: scrollDirection,
      itemCount: models.length,
      itemBuilder: (context, index) {
        return models[index].asContactAvatar(
          FFContactAvatarTheme.defaultTheme,
          onTap: () {
            onSelectAvatar(index);
          },
        );
      },
      itemExtent: theme.itemExtent,
      padding: theme.listPadding,
    );
  }

  @override
  Widget build(BuildContext context) {
    return ListView(
      scrollDirection: scrollDirection,
      itemExtent: theme.itemExtent,
      padding: theme.listPadding,
      children: models.map((model) {
        var index = models.indexOf(model);
        return model.asContactAvatar(
          theme,
          onTap: () {
            onSelectAvatar(index);
          },
        );
      }).toList(),
    );
  }
}


lib/ff-contact-avatar-theme.dart
import 'package:flutter/material.dart';

class FFContactAvatarTheme {
  final Color backgroundColor;
  final Color foregroundColor;
  final Color badgeColor;
  final double avatarRadius;
  final TextStyle initialsTextStyle;
  final TextStyle nameTextStyle;
  final TextStyle messageTextStyle;
  final double verticalPadding;
  final double avatarElevation;
  final double itemExtent;
  final EdgeInsets listPadding;

  FFContactAvatarTheme({
    this.backgroundColor = Colors.blueGrey,
    this.foregroundColor = Colors.white,
    this.badgeColor = Colors.red,
    this.avatarRadius = 31,
    this.initialsTextStyle = kInitialTextStyle,
    this.nameTextStyle = kNameTextStyle,
    this.messageTextStyle = kMessageTextStyle,
    this.verticalPadding = 2,
    this.avatarElevation = 4,
    this.itemExtent = 80,
    this.listPadding = kListPadding,
  });

  static const TextStyle kInitialTextStyle = TextStyle(
    fontWeight: FontWeight.bold,
    fontSize: 26,
    letterSpacing: 1.4,
  );

  static const TextStyle kNameTextStyle = TextStyle(
    fontWeight: FontWeight.bold,
    fontSize: 13.0,
  );

  static const TextStyle kMessageTextStyle = TextStyle(
    fontWeight: FontWeight.normal,
    fontSize: 11.0,
  );

  static const EdgeInsets kListPadding = EdgeInsets.all(16.0);

  static final FFContactAvatarTheme defaultTheme = FFContactAvatarTheme(
    backgroundColor: Colors.blueGrey,
    foregroundColor: Colors.white,
    badgeColor: Colors.red,
    avatarRadius: 31,
    initialsTextStyle: kInitialTextStyle,
    nameTextStyle: kNameTextStyle,
    messageTextStyle: kMessageTextStyle,
    verticalPadding: 2,
    avatarElevation: 4,
    itemExtent: 80,
    listPadding: kListPadding,
  );
}


lib/ff_contact_avatar_model.dart
import 'ff_contact_avatar.dart';
import 'package:flutter/material.dart';

class FFContactAvatarModel {
  final String name;
  final String message;
  final Image image;
  final bool showBadge;
  final VoidCallback onTap;

  FFContactAvatarModel({
    @required this.name,
    this.message,
    this.image,
    this.showBadge = false,
    this.onTap,
  });

  FFContactAvatar asContactAvatar(
    FFContactAvatarTheme theme, {
    VoidCallback onTap,
  }) {
    return (image == null)
        ? FFContactAvatar(
            name: this.name,
            message: this.message,
            showBadge: this.showBadge,
            theme: theme,
            onTap: onTap,
          )
        : FFContactAvatar(
            name: this.name,
            message: this.message,
            showBadge: this.showBadge,
            image: this.image,
            theme: theme,
            onTap: onTap,
          );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:ff_contact_avatar/ff_contact_avatar.dart';

import 'package:english_words/english_words.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'FFContactAvatar Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'FFContactAvatar Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  // Models is a short, manually created set of data, used to demonstrate
  // a ListView WITHOUT a ListView.builder
  final List<FFContactAvatarModel> models = [
    FFContactAvatarModel(
      name: 'Ed Roberts',
      message: 'He\s a man',
      showBadge: true,
    ),
    FFContactAvatarModel(
      name: 'Mr Ed',
      message: 'He\s a horse',
      showBadge: false,
      image: Image.network(
          'https://upload.wikimedia.org/wikipedia/commons/8/81/Mister_Ed.JPG'),
    ),
    FFContactAvatarModel(
      name: 'Mrs Roberts',
      message: 'Ed\'s Wife',
    ),
    FFContactAvatarModel(
      name: 'Mrs Horse',
      message: 'Also Ed\'s Wife',
      showBadge: true,
    ),
  ];

  // longList is populated with random data, used to demonstrate use WITH a ListView.builder
  final List<FFContactAvatarModel> longList = [];

  @override
  Widget build(BuildContext context) {
    // Create a 200 long list of data using random word pairs
    generateWordPairs().take(200).forEach((pair) {
      longList.add(
        FFContactAvatarModel(
          name: pair.first + ' ' + pair.second,
          message: generateWordPairs().take(1).first.join(' '),
        ),
      );
    });

    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            // --------- SECTION 1
            // Manually output 2 x avatars
            Text('Sample using Individual Avatars'),
            FFContactAvatar(
              name: 'Ed Roberts',
              message: 'He\'s a man',
              showBadge: true,
              theme: FFContactAvatarTheme(
                backgroundColor: Colors.blueAccent,
                avatarElevation: 20.0,
              ),
            ),
            SizedBox(height: 10),
            FFContactAvatar(
              name: 'Mr Ed',
              message: 'He\'s a horse',
              image: Image.network(
                  'https://upload.wikimedia.org/wikipedia/commons/8/81/Mister_Ed.JPG'),
              showBadge: true,
            ),
            SizedBox(
              height: 1,
              child: Container(
                color: Colors.grey,
              ),
            ),
            // --------------- SECTION 2
            // Output a ListView using the manual data set (NO builder)
            Text('Sample using full ListView'),
            Expanded(
              child: FFContactAvatarList.fromModelList(
                models,
                scrollDirection: Axis.horizontal,
                onSelectAvatar: (index) {
                  print('Index ' + index.toString() + ' selected.');
                },
              ),
            ),
            SizedBox(
              height: 1,
              child: Container(
                color: Colors.grey,
              ),
            ),
            // --------------- SECTION 3
            // Output a ListView using the .builder approach (using long list of generated data)
            Text('Sample using ListView.Builder'),
            Expanded(
              child: FFContactAvatarList.fromModelListWithBuilder(
                longList,
                context: context,
                scrollDirection: Axis.horizontal,
                onSelectAvatar: (index) {
                  print('Index ' + index.toString() + ' selected.');
                },
              ),
            ),
          ],
        ),
      ),
    );
  }
}
